File: aoc2025/go/challenges/day5.go
package challenges

import (
	"log"
	"log/slog"
	"sort"
	"strconv"
	"strings"

	"aoc/utilities"

	logger "github.com/Bparsons0904/goLogger"
)

type IngredientRange struct {
	min int
	max int
}

type Ingredients struct {
	FreshIngredientRanges []IngredientRange
	Ingredients           []int
}

func Day5() {
	log := logger.New("Day 5")

	file := utilities.ReadFile("day5.part1")

	var ingredients Ingredients
	ingredients.parseIngredients(file)

	timer := log.Timer("Part 1 Timer")
	part1Count, part2Count := ingredients.countFreshIngredients()
	timer()

	slog.Info("Part 1", "Part 1", part1Count, "Part 2", part2Count)
}

func (ingredients *Ingredients) countFreshIngredients() (int, int) {
	countOfFreshIngredients := 0
	for _, ingredient := range ingredients.Ingredients {
		for _, freshIngredientRange := range ingredients.FreshIngredientRanges {
			if freshIngredientRange.min <= ingredient && freshIngredientRange.max >= ingredient {
				countOfFreshIngredients++
				break
			}
		}
	}

	countOfPotentialFreshIngredients := 0
	for _, freshIngredientRange := range ingredients.FreshIngredientRanges {
		countOfPotentialFreshIngredients += (freshIngredientRange.max - freshIngredientRange.min) + 1
	}
	return countOfFreshIngredients, countOfPotentialFreshIngredients
}

func (ingredients *Ingredients) parseIngredients(file []string) {
	var ingredientRanges []IngredientRange
	isFirstHalf := true
	for _, row := range file {
		if row == "" {
			isFirstHalf = false
			continue
		}

		if isFirstHalf {
			ingredient := strings.Split(row, "-")
			min, err := strconv.Atoi(ingredient[0])
			if err != nil {
				log.Fatal(err)
			}

			max, err := strconv.Atoi(ingredient[1])
			if err != nil {
				log.Fatal(err)
			}
			ingredientRange := IngredientRange{
				min: min,
				max: max,
			}

			ingredientRanges = append(ingredientRanges, ingredientRange)
		} else {
			ingredientID, err := strconv.Atoi(row)
			if err != nil {
				log.Fatal(err)
			}

			ingredients.Ingredients = append(ingredients.Ingredients, ingredientID)
		}
	}

	sort.Slice(ingredientRanges, func(i, j int) bool {
		return ingredientRanges[i].min < ingredientRanges[j].min
	})

	for _, ingredientRange := range ingredientRanges {
		if len(ingredients.FreshIngredientRanges) == 0 {
			ingredients.FreshIngredientRanges = append(
				ingredients.FreshIngredientRanges,
				ingredientRange,
			)
			continue
		}
		found := false
		for i, freshIngredientRange := range ingredients.FreshIngredientRanges {
			if ingredientRange.min >= freshIngredientRange.min &&
				ingredientRange.min <= freshIngredientRange.max || (ingredientRange.max >= freshIngredientRange.min &&
				ingredientRange.max <= freshIngredientRange.max) {
				ingredients.FreshIngredientRanges[i].min = min(
					ingredients.FreshIngredientRanges[i].min,
					ingredientRange.min,
				)
				ingredients.FreshIngredientRanges[i].max = max(
					ingredients.FreshIngredientRanges[i].max,
					ingredientRange.max,
				)
				found = true
			}
		}
		if !found {
			ingredients.FreshIngredientRanges = append(
				ingredients.FreshIngredientRanges,
				ingredientRange,
			)
		}
	}
}

File: aoc2025/rust/src/bin/day5.rs
use std::cmp::{max};
use std::fs::File;
use std::io::{self, BufRead};
use std::time::Instant;

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
struct IngredientRange {
    min: i64,
    max: i64,
}

struct Ingredients {
    fresh_ingredient_ranges: Vec<IngredientRange>,
    ingredients: Vec<i64>,
}

fn main() {
    let ingredients = parse_ingredients("files/day5.part1").unwrap();
    let now = Instant::now();
    let (part1_count, part2_count) = count_fresh_ingredients(&ingredients);
    let elapsed = now.elapsed();

    println!("Day 5, Part 1: {}", part1_count);
    println!("Day 5, Part 2: {}", part2_count);
    println!("Part 1 & 2 took: {:.2?}", elapsed);
}

fn count_fresh_ingredients(ingredients: &Ingredients) -> (i64, i64) {
    let mut count_of_fresh_ingredients = 0;
    for &ingredient in &ingredients.ingredients {
        for range in &ingredients.fresh_ingredient_ranges {
            if range.min <= ingredient && range.max >= ingredient {
                count_of_fresh_ingredients += 1;
                break;
            }
        }
    }

    let mut count_of_potential_fresh_ingredients = 0;
    for range in &ingredients.fresh_ingredient_ranges {
        count_of_potential_fresh_ingredients += (range.max - range.min) + 1;
    }

    (count_of_fresh_ingredients, count_of_potential_fresh_ingredients)
}

fn parse_ingredients(filename: &str) -> io::Result<Ingredients> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);

    let mut ingredient_ranges = Vec::new();
    let mut ingredients = Vec::new();
    let mut is_first_half = true;

    for line in reader.lines() {
        let line = line?;
        if line.is_empty() {
            is_first_half = false;
            continue;
        }

        if is_first_half {
            let parts: Vec<&str> = line.split('-').collect();
            if parts.len() == 2 {
                let min = parts[0].parse::<i64>().unwrap();
                let max = parts[1].parse::<i64>().unwrap();
                ingredient_ranges.push(IngredientRange { min, max });
            }
        } else {
            let ingredient_id = line.parse::<i64>().unwrap();
            ingredients.push(ingredient_id);
        }
    }

    ingredient_ranges.sort_by_key(|r| r.min);

    let mut fresh_ingredient_ranges = Vec::new();
    if !ingredient_ranges.is_empty() {
        let mut current_range = ingredient_ranges[0];
        for &next_range in &ingredient_ranges[1..] {
            if next_range.min <= current_range.max + 1 {
                current_range.max = max(current_range.max, next_range.max);
            } else {
                fresh_ingredient_ranges.push(current_range);
                current_range = next_range;
            }
        }
        fresh_ingredient_ranges.push(current_range);
    }
    
    Ok(Ingredients { fresh_ingredient_ranges, ingredients })
}